contracts/common/GetActionSummary.sol

```diff
@@ -0,0 +1,203 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { IColony } from "./../colony/IColony.sol";
+import { ExtractCallData } from "./ExtractCallData.sol";
+import { GetActionDomainSkillId } from "./GetActionDomainSkillId.sol";
+import { ColonyDataTypes } from "./../colony/ColonyDataTypes.sol";
+import { ColonyRoles } from "./../colony/ColonyRoles.sol";
+import { Bytes4Includes } from "./Bytes4Includes.sol";
+
+pragma solidity 0.8.25;
+pragma experimental ABIEncoderV2;
+
+struct ActionSummary {
+  bytes4 sig;
+  uint256 domainSkillId;
+  uint256 expenditureId;
+  bytes32 requiredPermissions;
+}
+
+contract GetActionSummary is ExtractCallData, GetActionDomainSkillId {
+  using Bytes4Includes for bytes4[];
+
+  bytes4 constant MULTICALL = bytes4(keccak256("multicall(bytes[])"));
+  bytes4 constant SET_EXPENDITURE_STATE =
+    bytes4(
+      keccak256("setExpenditureState(uint256,uint256,uint256,uint256,bool[],bytes32[],bytes32)")
+    );
+  bytes4 constant SET_EXPENDITURE_PAYOUT =
+    bytes4(keccak256("setExpenditurePayout(uint256,uint256,uint256,uint256,address,uint256)"));
+
+  bytes32 constant ROOT_ROLES = ((bytes32(uint256(1)) <<
+    uint8(ColonyDataTypes.ColonyRole.Recovery)) |
+    (bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Root)));
+
+  bytes32 constant ONLY_ROOT_ROLE_MASK =
+    bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Root);
+
+  bytes4 constant SET_USER_ROLES =
+    bytes4(keccak256("setUserRoles(uint256,uint256,address,uint256,bytes32)"));
+
+  bytes32 constant ONLY_ARCHITECTURE_ROLE_MASK =
+    bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Architecture);
+
+  function getExpenditureId(bytes memory action) internal pure returns (uint256 expenditureId) {
+    bytes4 sig = getSig(action);
+    assert(isExpenditureSig(sig));
+
+    assembly {
+      expenditureId := mload(add(action, 0x64))
+    }
+  }
+
+  function getSingleActionSummary(
+    address colonyNetworkAddress,
+    address colonyAddress,
+    bytes memory _action,
+    address _altTarget
+  ) public view returns (ActionSummary memory) {
+    bytes4 sig = getSig(_action);
+    require(sig != MULTICALL, "colony-get-action-summary-no-nested-multicalls");
+
+    // TODO: explicitly check `isExtension` for target, currently this simply errors
+    address target = getTarget(_altTarget, colonyAddress);
+    bytes32 permissionMask = ColonyRoles(target).getCapabilityRoles(sig);
+
+    ActionSummary memory summary;
+    summary.sig = sig;
+
+    if (isExpenditureSig(sig)) {
+      summary.domainSkillId = getActionDomainSkillId(_action, colonyNetworkAddress, colonyAddress);
+      summary.expenditureId = getExpenditureId(_action);
+    } else if (permissionMask | ROOT_ROLES == ROOT_ROLES) {
```

Comment:
Can you add a comment clarifying the purpose of this branch? It's escaping me -- a function with no permission mask is being assigned the root role.

Reply:
Given it a crack. It's not quite as simple as I made the comments sound though - you need the arguments to whatever function being called (e.g. `setExpenditureMetadata` to also be valid arguments to `getActionDomainSkillId` and still have everything match.

Hopefully the motivation behind the block is clear, though, even if I haven't been able to write the test 😅